### PR TITLE
[#5145] improvement(flink-connector): Shrink the flink-connector-runtime lib size from 84MB to 16MB

### PR DESCRIPTION
**Pull request by coolderli**
### What changes were proposed in this pull request?

- shrink the flink-connector-runtime jar.
- Change client-java scope from implementation to CompileOnly  in flink module.
- use the client-java-runtime in flink-connector-runtime.

### Why are the changes needed?
Fix: #5145 

- The flink-connector-runtime jar is much big as mentioned in https://github.com/apache/gravitino/pull/4791#discussion_r1738193021.

### Does this PR introduce _any_ user-facing change?

- no

### How was this patch tested?

- local test

### CODE REVIEW

**Comment by coolderli** [comment]
image

**Comment by coolderli** [comment]
the dependencies of flink-connector-runtime:

```
implementation - Implementation dependencies for the 'main' feature. (n)
+--- project client-java-runtime (n)
\--- project flink (n)

runtimeClasspath - Runtime classpath of source set 'main'.
+--- project :clients:client-java-runtime
\--- project :flink-connector:flink


```

File: flink-connector/flink-runtime/build.gradle.kts [hunk 1]
```diff
@@ -56,7 +56,6 @@ tasks.withType<ShadowJar>(ShadowJar::class.java) {
   relocate("com.google", "org.apache.gravitino.shaded.com.google")
   relocate("google", "org.apache.gravitino.shaded.google")
   relocate("org.apache.hc", "org.apache.gravitino.shaded.org.apache.hc")
-  relocate("org.apache.commons", "org.apache.gravitino.shaded.org.apache.commons")
```
Comment (coolderli) [inline flink-connector/flink-runtime/build.gradle.kts:59]: remove it as it is already shaded in client-java-runtime.

**Comment by coolderli** [comment]
@FANNG1 @mchades PTAL. Thanks.

File: flink-connector/flink/build.gradle.kts [hunk 2]
```diff
@@ -38,19 +38,21 @@ val scalaVersion: String = "2.12"
...
-  implementation(libs.guava)
-  implementation(libs.httpclient5)
-  implementation(libs.jackson.databind)
-  implementation(libs.jackson.annotations)
-  implementation(libs.jackson.datatype.jdk8)
-  implementation(libs.jackson.datatype.jsr310)
+  compileOnly(libs.commons.lang3)
+  compileOnly(libs.guava)
+  compileOnly(libs.httpclient5)
+  compileOnly(libs.jackson.databind)
```
Comment (FANNG1) [inline flink-connector/flink/build.gradle.kts:52]: why using jason in Flink?

File: flink-connector/flink/build.gradle.kts [hunk 3]
```diff
@@ -38,19 +38,15 @@ val scalaVersion: String = "2.12"
...
 
 dependencies {
-  implementation(project(":api"))
-  implementation(project(":catalogs:catalog-common"))
-  implementation(project(":common"))
-  implementation(project(":core"))
-
+  // use the client-java instead of client-java-runtime to provide the conflict from
+  // the shaded of com.fasterxml.
+  compileOnly(project(":clients:client-java"))
```
Comment (mchades) [inline flink-connector/flink/build.gradle.kts:43]: Please make those in alphabetical order.

**Comment by jerryshao** [comment]
Do we need to merge this to 0.6.0?

File: flink-connector/flink/build.gradle.kts (hunk 3 above)
Comment (jerryshao) [reply flink-connector/flink/build.gradle.kts:43]: Why do we use client-java, not client-java-runtime, the comment doesn't explain clearly.

Comment (jerryshao) [reply flink-connector/flink/build.gradle.kts:43]: Also, why the scope is "compileOnly"?

**Comment by FANNG1** [comment]
> Do we need to merge this to 0.6.0?

It's an improvement, I prefer not to merge in 0.6.0

**Comment by jerryshao** [comment]
@coolderli can you please continue on this, thanks.

**Comment by coolderli** [comment]
@FANNG1 @mchades Please help review this, Thanks.

**Comment by FANNG1** [comment]
Could you update the PR title to add the shrink info like `shrink from xxM to xxM`?  and paste the related dependence jars in the comment like https://github.com/apache/gravitino/pull/4548?

**Comment by coolderli** [comment]
@FANNG1 
I can't find a way to print the dependent jar in the connector jar. Here are the dependencies.
```
------------------------------------------------------------
Project ':flink-connector:flink'
------------------------------------------------------------

runtimeClasspath - Runtime classpath of source set 'main'.
+--- project :catalogs:catalog-common
|    +--- org.apache.commons:commons-lang3:3.14.0
|    +--- com.google.guava:guava:32.1.3-jre
|    |    +--- com.google.guava:failureaccess:1.0.1
|    |    +--- com.google.guava:listenablefuture:9999.0-empty-to-avoid-conflict-with-guava
|    |    +--- com.google.code.findbugs:jsr305:3.0.2
|    |    +--- org.checkerframework:checker-qual:3.37.0
|    |    \--- com.google.errorprone:error_prone_annotations:2.21.1
|    \--- org.slf4j:slf4j-api:2.0.9
\--- com.google.guava:guava:32.1.3-jre (*)

------------------------------------------------------------
Project ':flink-connector:flink-runtime'
------------------------------------------------------------

implementation - Implementation dependencies for the 'main' feature. (n)
+--- project client-java-runtime (n)
\--- project flink (n)
```

**Comment by coolderli** [comment]
@FANNG1 Hi, here are the dependency jars. Please take a look.
```
total 16M
-rw-rw-r-- 1 mi mi  41K 10月 15 14:57 gravitino-flink-1.18_2.12-0.7.0-incubating-SNAPSHOT.jar
-rw-rw-r-- 1 mi mi  12M 10月 15 14:57 gravitino-client-java-runtime-0.7.0-incubating-SNAPSHOT.jar
-rw-rw-r-- 1 mi mi  14K 10月 15 14:57 gravitino-catalog-common-0.7.0-incubating-SNAPSHOT.jar
-rw-rw-r-- 1 mi mi  64K 10月 15 14:57 slf4j-api-2.0.9.jar
-rw-rw-r-- 1 mi mi 2.2K 10月 15 14:57 listenablefuture-9999.0-empty-to-avoid-conflict-with-guava.jar
-rw-rw-r-- 1 mi mi  20K 10月 15 14:57 jsr305-3.0.2.jar
-rw-rw-r-- 1 mi mi 3.0M 10月 15 14:57 guava-32.1.3-jre.jar
-rw-rw-r-- 1 mi mi 4.6K 10月 15 14:57 failureaccess-1.0.1.jar
-rw-rw-r-- 1 mi mi  17K 10月 15 14:57 error_prone_annotations-2.21.1.jar
-rw-rw-r-- 1 mi mi 643K 10月 15 14:57 commons-lang3-3.14.0.jar
-rw-rw-r-- 1 mi mi 220K 10月 15 14:57 checker-qual-3.37.0.jar

```

File: flink-connector/flink-runtime/build.gradle.kts [hunk 4]
```diff
@@ -67,6 +66,11 @@ publishing {
   }
 }
 
+tasks.register<Copy>("copy") {
```
Comment (FANNG1) [inline flink-connector/flink-runtime/build.gradle.kts:69]: please remove this

**Comment by FANNG1** [comment]
LGTM except one comment

**Comment by mchades** [comment]
This is not a minor change, could you plz create a new subtask issue of #4513 to record this?

**Comment by coolderli** [comment]
@FANNG1 @mchades I have addressed the comment. Please take a look. Thanks.

**Review by mchades — COMMENTED** [review]
LGTM.

**Review by FANNG1 — APPROVED** [review]

**Pull request merged**